lib/async-emitter.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+class AsyncEmitter {
+  constructor() {
+    this.events = new Map();
+  }
+
+  // Get or create event
+  //   name <string> event name
+  // Returns: { on: Set { <Function> }, once: Set{ <Function> } }
+  event(name) {
+    const { events } = this;
+    const event = events.get(name);
+    if (event) return event;
+    const res = { on: new Set(), once: new Set() };
+    events.set(name, res);
+    return res;
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  on(name, fn) {
+    this.event(name).on.add(fn);
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  // Returns: <Promise> | <null>
+  once(name, fn) {
+    if (fn === undefined) {
+      return new Promise(resolve => {
+        this.once(name, resolve);
+      });
+    }
+    this.event(name).once.add(fn);
+    return null;
+  }
+
+  // Emit event
+  //   name <string> event name
+  //   args <any[]>
+  // Returns: <Promise> | <null>
+  emit(name, ...args) {
+    const { events } = this;
+    const event = events.get(name);
+    if (!event) return null;
```

Comment:
I'd prefer to return `Promise.resolve([])` (empty array to be consistent with what the other branch returns - an array of resolved values) here to allow to always just `.then` on a return value without checking if the null was returned.

Reply:
I think `emit('name')` will be called only with `await` like this: `await emit('name')` so 'null' will be ok.

Reply:
@tshemsedinov and in case it is not called with `await`?
Why do users have to wrap the result in `Promise.resolve()` every time because there is no guarantee that the function will return a Promise? Why can't we just make this function `async`?

Reply:
@belochub Why user need `Promise.resolve()`? `await null` here works like `await Promise.resolve(null)`. I mean users can call `emit('name')` if they want not to wait or `await emit('name')` if they want to wait but why somebody will use `Promise.resolve(ae.emit('name'))` ?

Reply:
@tshemsedinov As I said
>@tshemsedinov and in case it is not called with await?

That would require to either check the returned value or wrap the result in `Promise.resolve()` before calling `.then()`, both of which isn't very convenient. Once again, I don't see why we can't make this method an `AsyncFunction`.